src/Workspaces/Remote/ServiceHub/Host/RemoteWorkspace.SolutionCreator.cs

```diff
@@ -441,6 +446,24 @@ private async Task<Project> UpdateProjectInfoAsync(Project project, Checksum inf
                 }
 
                 return project;
+
+                async Task PopulateOldDocumentMapAsync()
+                {
+                    foreach (var (_, state) in existingTextDocumentStates.States)
+                    {
+                        var documentChecksums = await state.GetStateChecksumsAsync(cancellationToken).ConfigureAwait(false);
+                        if (olds.Object.Contains(documentChecksums.Checksum))
+                            oldMap.Add(state.Id, documentChecksums);
+                    }
+                }
+
+                async Task PopulateNewDocumentMapAsync(SolutionCreator @this)
+                {
+                    var documentStateChecksums = await @this._assetProvider.GetAssetsAsync<DocumentStateChecksums>(news.Object, cancellationToken).ConfigureAwait(false);
+
+                    foreach (var (_, documentStateChecksum) in documentStateChecksums)
+                        newMap.Add(documentStateChecksum.DocumentId, documentStateChecksum);
```

Comment:
this is much more efficient than before.  similar to projects, this avoids an unnecessary round trip to determine document ids.